api/trace/b3_propagator_integration_test.go

```diff
@@ -0,0 +1,290 @@
+	SpanIDHeader       string
+	SampledHeader      string
+	ParentSpanIDHeader string
+}
+
+func (s *Supplier) Get(key string) string {
+	switch key {
+	case trace.B3SingleHeader:
+		return s.SingleHeader
+	case trace.B3DebugFlagHeader:
+		return s.DebugFlagHeader
+	case trace.B3TraceIDHeader:
+		return s.TraceIDHeader
+	case trace.B3SpanIDHeader:
+		return s.SpanIDHeader
+	case trace.B3SampledHeader:
+		return s.SampledHeader
+	case trace.B3ParentSpanIDHeader:
+		return s.ParentSpanIDHeader
+	}
+	return ""
+}
+
+func (s *Supplier) Set(key, value string) {
+	fmt.Println("called", key, value)
+	switch key {
+	case trace.B3SingleHeader:
+		s.SingleHeader = value
+	case trace.B3DebugFlagHeader:
+		s.DebugFlagHeader = value
+	case trace.B3TraceIDHeader:
+		s.TraceIDHeader = value
+	case trace.B3SpanIDHeader:
+		s.SpanIDHeader = value
+	case trace.B3SampledHeader:
+		s.SampledHeader = value
+	case trace.B3ParentSpanIDHeader:
+		s.ParentSpanIDHeader = value
+	}
+}
+
+type TestSpan struct {
+	trace.NoopSpan
+	sc trace.SpanContext
+}
+
+func (s TestSpan) SpanContext() trace.SpanContext {
+	return s.sc
+}
+
+func TestInject(t *testing.T) {
+	tests := []struct {
+		sc       trace.SpanContext
+		b3       trace.B3
+		expected *Supplier
+	}{
+		{
+			sc:       trace.SpanContext{},
+			b3:       trace.B3{},
+			expected: &Supplier{},
+		},
+		{
+			sc: trace.SpanContext{
+				SpanID:     spanID,
+				TraceFlags: trace.FlagsSampled,
+			},
+			b3:       trace.B3{},
+			expected: &Supplier{},
+		},
+		{
+			sc: trace.SpanContext{
+				TraceID:    traceID,
+				TraceFlags: trace.FlagsSampled,
+			},
+			b3:       trace.B3{},
+			expected: &Supplier{},
+		},
+		{
+			sc: trace.SpanContext{
+				TraceFlags: trace.FlagsSampled,
+			},
+			b3:       trace.B3{},
+			expected: &Supplier{},
+		},
+		{
+			sc: trace.SpanContext{
+				TraceID:    traceID,
+				SpanID:     spanID,
+				TraceFlags: trace.FlagsSampled,
+			},
+			b3: trace.B3{SingleHeader: true},
+			expected: &Supplier{
+				SingleHeader: "000000000000007b00000000000001c8-000000000000007b-1",
+			},
+		},
+		{
+			sc: trace.SpanContext{
+				TraceID: traceID,
+				SpanID:  spanID,
+			},
+			b3: trace.B3{SingleHeader: true},
+			expected: &Supplier{
+				SingleHeader: "000000000000007b00000000000001c8-000000000000007b-0",
+			},
+		},
+		{
+			sc: trace.SpanContext{
+				TraceID:    traceID,
+				SpanID:     spanID,
+				TraceFlags: trace.FlagsUnused,
+			},
+			b3: trace.B3{SingleHeader: true},
+			expected: &Supplier{
+				SingleHeader: "000000000000007b00000000000001c8-000000000000007b",
+			},
+		},
+		{
+			sc: trace.SpanContext{
+				TraceID:    traceID,
+				SpanID:     spanID,
+				TraceFlags: trace.FlagsSampled,
+			},
+			b3: trace.B3{SingleHeader: true, SingleAndMultiHeader: true},
+			expected: &Supplier{
+				TraceIDHeader: traceIDStr,
+				SpanIDHeader:  spanIDStr,
+				SampledHeader: "1",
+				SingleHeader:  "000000000000007b00000000000001c8-000000000000007b-1",
+			},
+		},
```

Comment:
Does it make sense to add a case for `b3: trace.B3{SingleAndMultiHeader: true}`, i.e. where it overrides `SingleHeader: false`?

Reply:
Yeah, good catch. I'll add that.

FWIW, I'm refactoring these to integrate with the tests in https://github.com/open-telemetry/opentelemetry-go/blob/master/api/trace/testtrace/b3_propagator_test.go. I'll be sure to include that testcase.